### PR TITLE
AArch64: Implement JIT field watch support

### PR DESCRIPTION
**Pull request by Akira1Saitoh**
This commit adds support for JIT field watch on aarch64.

Depends on https://github.com/eclipse/omr/pull/6252

Signed-off-by: Akira Saitoh <saiaki@jp.ibm.com>

### CODE REVIEW

File: runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp [hunk 1]
```diff
@@ -103,16 +105,367 @@ void VMgenerateCatchBlockBBStartPrologue(TR::Node *node, TR::Instruction *fenceI
...
+ * @param[in] dt: data type
+ * @param[in] elementSize: element size
+ * @param[in] isLoad : true if operation is load
+ * @return opcode
+ */
+static TR::InstOpCode::Mnemonic getLoadOrStoreFromDataType(TR::CodeGenerator *cg, TR::DataType dt, int32_t elementSize, bool isLoad)
+   {
+   switch (dt)
+      {
+   case TR::Int8:
```
Comment (knn-k) [inline runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp:137]: Please change the indentation of case labels in this switch statement like this.
```
   switch (dt)
      {
      case TR::Int8:
         return isLoad ? TR::InstOpCode::ldrbimm : TR::InstOpCode::strbimm;
      case TR::Int32:
         ...
         return TR::InstOpCode::bad;
      }
```

File: runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp [hunk 2]
```diff
@@ -103,16 +105,367 @@ void VMgenerateCatchBlockBBStartPrologue(TR::Node *node, TR::Instruction *fenceI
...
+      return isLoad ? TR::InstOpCode::ldrbimm : TR::InstOpCode::strbimm;
+   case TR::Int16:
+      return isLoad ? TR::InstOpCode::ldrhimm : TR::InstOpCode::strhimm;
```
Comment (knn-k) [inline runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp:140]: The Int8 and Int16 cases will generate unsigned instructions for the `isLoad` case.
Don't you need to take signed cases into consideration?

This function (`getLoadOrStoreFromDataType()`) is called only for storing case in this PR.
I think you can use the existing function `getStoreOpCodeFromDataType()`.

Comment (Akira1Saitoh) [reply runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp:140]: Ah, I did not notice the existing function. I will change the code to use existing one.

File: runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp [hunk 3]
```diff
@@ -103,16 +105,367 @@ void VMgenerateCatchBlockBBStartPrologue(TR::Node *node, TR::Instruction *fenceI
...
+
+   generateReportOOL->swapInstructionListsWithCompilation();
+
+   generateLabelInstruction(cg, TR::InstOpCode::label, node, endLabel);
+
+   cg->stopUsingRegister(scratchReg);
+   if (!isSideEffectReg)
+      cg->stopUsingRegister(fieldClassReg);
+
+   cg->machine()->setLinkRegisterKilled(true);
```
Comment (knn-k) [inline runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp:345]: I prefer this `setLinkRegisterKilled(true)` to be in `generateReportFieldAccessOutlinedInstructions()`, after generaing the `bl` instruction.

File: runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp [hunk 4]
```diff
@@ -103,16 +105,367 @@ void VMgenerateCatchBlockBBStartPrologue(TR::Node *node, TR::Instruction *fenceI
...
+      }
+
+   TR::MemoryReference *classFlagsMemRef = new (cg->trHeapMemory()) TR::MemoryReference(fieldClassReg, static_cast<int32_t>(fej9->getOffsetOfClassFlags()), cg);
+
+   generateTrg1MemInstruction(cg, TR::InstOpCode::ldrimmw, node, scratchReg, classFlagsMemRef);
+   static_assert(J9ClassHasWatchedFields == 0x100, "We assume that J9ClassHasWatchedFields is 0x100");
+   generateTestImmInstruction(cg, node, scratchReg, 0x600); // 0x600 is immr:imms for 0x100
+   generateConditionalBranchInstruction(cg, TR::InstOpCode::b_cond, node, fieldReportLabel, TR::CC_NE);
+
+	generateReportOOL->swapInstructionListsWithCompilation();
```
Comment (knn-k) [inline runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp:332]: This line has a different indentation.

File: runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp (hunk 3 above)
Comment (Akira1Saitoh) [reply runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp]: Moved `setLinkRegisterKilled(true)` into `generateReportFieldAccessOutlinedInstructions()`.

File: runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp (hunk 4 above)
Comment (Akira1Saitoh) [reply runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp]: Fixed it.

**Comment by knn-k** [comment]
jenkins test sanity alinux64 jdk11 depends eclipse/omr#6252

**Comment by knn-k** [comment]
jenkins test extended.functional alinux64 jdk11 depends eclipse/omr#6252

**Comment by knn-k** [comment]
jenkins test sanity alinux64 jdk11 depends eclipse/omr#6252

**Review by knn-k — APPROVED** [review]

**Pull request merged**